ocfweb/docs/views/servers.py

```diff
@@ -56,96 +60,96 @@ def english_type(self):
-            hostname='riptide',
-            type='hypervisor',
-            children=[
-                Host.from_ldap(hostname)
-                for hostname in (
-                    'anthrax',
-                    'biohazard',
-                    'death',
-                    'dementors',
-                    'democracy',
-                    'firestorm',
-                    'flood',
-                    'fraud',
-                    'gridlock',
-                    'lethe',
-                    'lightning',
-                    'maelstrom',
-                    'monsoon',
-                    'pestilence',
-                    'reaper',
-                    'segfault',
-                    'supernova',
-                    'thunder',
-                    'tsunami',
-                    'vampires',
-                    'whiteout',
-                    'zombies',
-                )
-            ],
-        ),
 
-        Host.from_ldap(
-            hostname='corruption',
-            type='server',
-        ),
+PQL_GET_VMS = "facts { name = 'vms' }"
+PQL_IS_HYPERVISOR = 'resources[certname] { type = "Class" and title = "Ocf_kvm" }'
+
 
-        Host.from_ldap(
-            hostname='dev-fallingrocks',
-            type='server',
+def query_puppet(query):
+    """Accepts a PQL query, returns a parsed json result."""
+    r = requests.get(
+        PUPPETDB_URL,
+        cert=(
+            os.path.join(PUPPET_CERT_DIR, 'puppet-cert.pem'),
+            os.path.join(PUPPET_CERT_DIR, 'puppet-private.pem'),
         ),
+        verify=os.path.join(PUPPET_CERT_DIR, 'puppet-ca.pem'),
+        params={'query': query},
+    )
+    return r.json() if r.status_code == 200 else None
+
+
+def format_query_output(item):
+    """Converts an item of a puppet query to tuple(hostname, query_value)."""
+    return item['certname'].split('.')[0], item.get('value')
+
 
-        Host('blackhole', 'network', 'Managed Arista 7050SX-64 Switch.', []),
+def ldap_to_host(item):
+    """Accepts an ldap output item, returns tuple(hostname, host_object)."""
+    description = item.get('description', [''])[0]
+    hostname = item['cn'][0]
+    return hostname, Host(hostname, item['type'], description, ())
 
-        Host('logjam', 'printer', '', []),
-        Host('pagefault', 'printer', '', []),
-        Host('papercut', 'printer', '', []),
-        Host.from_ldap('overheat', type='raspi'),
-        Host.from_ldap('tornado', type='nuc'),
 
-        *(Host.from_ldap(desktop, type='desktop')
-          for desktop in sorted(list_desktops())),
-    ]
+@periodic(300)
+def get_hosts():
+    ldap_output = hosts_by_filter('(|(type=server)(type=desktop)(type=printer))')
+    servers = dict(ldap_to_host(item) for item in ldap_output if not is_hidden(item))
+
+    hypervisors_hostnames = dict(format_query_output(item) for item in query_puppet(PQL_IS_HYPERVISOR))
+    all_children = dict(format_query_output(item) for item in query_puppet(PQL_GET_VMS))
+
+    hostnames_seen = []
```

Comment:
I would use a set for this variable:

```python
hostnames_seen = set()
```

A set allows you to test membership in constant time instead of linear time.